Add endpoint for 2 raster pair counting

mmcfarland opened the pull request:
#### Overview

Builds on #7 

To implement a 2 raster operation, similar to that used in MMW TR-55 implementation, this PR adds an endpoint which counts the unique pairs of cell values occurring when two rasters are stacked on each other.
- Took a deep dive into numpy methods with much assistance from stackoverflow and the docs.  As this is a research project, I tried to comment everything copiously, though some of the lines could easily be condensed into a terser form otherwise.
- Extracted out some common elements between the two existing functions.  There's a bit more that is repeated, but I'm anticipating some future work that might appreciate being able to modify those bits.
- A note on performance: while a large polygon covering all of Philadelpia and Camden takes ~0.18 seconds locally, increasing to a much larger polygon (~ 1/5 of PA) can take up to 20 seconds.  The execution time doesn't increase on the same scale as the existing `/counts` endpoint, but I have't profiled to see what is the most expensive operations here.
#### Testing

Ideally, two rasters would be used to test this endpoint, but they must be in the same shape and SRS.  Using the same raster twice actually helps evaluate the correctness.  If you give the same raster path as both rasters, then all pairs counted should have identical values from both rasters, ie, no mixed values are possible (`11:11` vs `11:34`) 

Sample HTTP body to post to `/pair-counts`:

``` json
{
    "rasters": ["nlcd/nlcd_2011_landcover_2011_edition_2014_10_10.img", "nlcd/nlcd_2011_landcover_2011_edition_2014_10_10.img"],
    "queryPolygon": {
        "type": "Polygon",
        "coordinates": [
          [
            [
              -75.26870727539062,
              39.876546372401194
            ],
            [
              -75.26870727539062,
              40.05127080263306
            ],
            [
              -75.03868103027344,
              40.05127080263306
            ],
            [
              -75.03868103027344,
              39.876546372401194
            ],
            [
              -75.26870727539062,
              39.876546372401194
            ]
          ]
        ]
     }
}
```

returns:

``` json
{
  "pairs": {
    "11::11": 31371,
    "21::21": 55497,
    "22::22": 64413,
    "23::23": 125696,
    "24::24": 114420,
    "31::31": 380,
    "41::41": 18562,
    "42::42": 127,
    "43::43": 1102,
    "52::52": 1206,
    "71::71": 752,
    "81::81": 1240,
    "82::82": 503,
    "90::90": 5675,
    "95::95": 4255
  },
  "time": 0.20161300000000004
}
```

mmcfarland committed:
Implement a 2-raster pair counting endpoint

First attempt at a TR55 like 2-raster cell pairing.  Works, but is not
fast.

mmcfarland committed:
Improve performce of pair counts

Significantly speeds up 2 raster element-wise pairing using numpy
methods as observed here:
http://stackoverflow.com/questions/30879446/efficiently-count-the-number-of-occurrences-of-unique-subarrays-in-numpy

Commit:
Merge pull request #2 from azavea/feature/mjm/add-counts

Add counts

mmcfarland committed:
Extract vector masking function

Extract the function which masks a reaster by a vector shape out to a
well documented common utility function.

hectcastro commented on src/geop/main.py:
```diff
@@ -66,6 +44,66 @@ def count():
+    trim_arr = np.ma.compress_rowcols(arr, 0)
+
+    # Lexicographically sort so that repeated pairs follow one another
+    sorted_arr = trim_arr[np.lexsort(trim_arr.T), :]
```
In profiling, the `np.lexsort` stood out just a little bit more than other operations in terms of execution time. I think we'd have to profile it with a larger polygon (the 1/5 PA you mentioned in the PR description maybe) to see how much more/less it stands out.

hectcastro commented on src/geop/main.py:
```diff
@@ -66,6 +44,66 @@ def count():
+    sorted_arr = trim_arr[np.lexsort(trim_arr.T), :]
+
+    # Get the indices where a new row of pairs appears
+    diff_indexes = np.where(np.any(np.diff(sorted_arr, axis=0), 1))[0]
```
Is `np.diff` causing like cells in both rasters to negate themselves in a way that only produces entries in `diff_indexes` for indexes (after `np.any` and `np.where` are applied) that aren't zeroed out?

Commit:
Merge pull request #7 from azavea/feature/mjm/config-req

Move request info to JSON config

Commit:
Merge pull request #13 from azavea/feature/mjm/geo-query-counts

Feature/mjm/geo query counts

Commit:
Correct key name in error message

The error was giving the python variable name, not the JSON key name
actually expected.

Commit:
Extract out nested numpy calls

For better clarity and profiling, move a nested numpy stack of calls to
single line operations.
